WordPress/openverse
# Add a Slack notification job to the CI + CD workflow

## Fixes




Fixes #1064 by @zackkrida 

## Description




This PR adds a job to the CI + CD workflow that runs after the docs are built, Docker images are published and staging sites are deployed. It sends a Slack report if any of these steps have been skipped or failed.

## Testing Instructions

Read carefully, and hope it works. You can extract the Python script out to a file and run it against your own env vars to see the output.

```console
$ env \
    SERVER_URL=https://github.com \
    REPOSITORY=WordPress/openverse \
    RUN_ID=4556856617 \
    EMIT_DOCS_RESULT=success \
    PUBLISH_IMAGES_RESULT=failure \
    DEPLOY_FRONTEND_RESULT=skipped \
    DEPLOY_API_RESULT=cancelled \
    GITHUB_OUTPUT=/dev/null \
  python ping.py
```

```
payload={"text": "1 success, 1 failure, 1 skipped, 1 cancelled", "blocks": [{"type": "section", "text": {"type": "mrkdwn", "text": "<https://github.com/WordPress/openverse/actions/runs/4556856617|Click here to review the completed CI + CD workflow>."}}, {"type": "section", "fields": [{"type": "mrkdwn", "text": "*emit-docs:*\n:tadaco: success"}, {"type": "mrkdwn", "text": "*publish-images:*\n:alert: failure"}, {"type": "mrkdwn", "text": "*deploy-frontend:*\n:warning: skipped"}, {"type": "mrkdwn", "text": "*deploy-api:*\n:no-good: cancelled"}]}]}
```

```console
$ curl \
  -X POST \
  -H 'Content-type: application/json' \
  --data '{"text": "1 success, 1 failure, 1 skipped, 1 cancelled", "blocks": [{"type": "section", "text": {"type": "mrkdwn", "text": "<https://github.com/WordPress/openverse/actions/runs/4556856617|Click here to review the completed CI + CD workflow>."}}, {"type": "section", "fields": [{"type": "mrkdwn", "text": "*emit-docs:*\n:workflow-success: success"}, {"type": "mrkdwn", "text": "*publish-images:*\n:workflow-failure: failure"}, {"type": "mrkdwn", "text": "*deploy-frontend:*\n:workflow-skipped: skipped"}, {"type": "mrkdwn", "text": "*deploy-api:*\n:workflow-cancelled: cancelled"}]}]}' \
  <slack webhook url>
```

<img width="442" alt="Screenshot 2023-03-30 at 12 58 22 AM" src="https://user-images.githubusercontent.com/16580576/228665911-66643518-60ad-4280-a70c-ca9471a63569.png">

## Checklist



      `Update index.md`).
      a parent feature branch.
      errors.

[best_practices]:
  https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin



<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>
